### PR TITLE
[WIP] Migration Phase 2

### PR DESCRIPTION
**Pull request by kaderson1**

### CODE REVIEW

File: src/Configuration.ts [hunk 1]
```diff
@@ -70,6 +91,7 @@ export type IMigrationParametersConfig = t.TypeOf<
 
 export const IConfiguration = t.interface({
   SOURCE: IEnvironmentConfig,
+  TARGET: IEnvironmentConfig,
```
Comment (CarmenManzulli) [inline src/Configuration.ts]: Make IEnvironmentConfig partial using t.intersection

File: src/utils/WatsonUtils.ts [hunk 2]
```diff
@@ -103,6 +104,20 @@ export function getWorkspaceInformationById(
   });
 }
 
+// upload workspace single or all workspaces
```
Comment (CarmenManzulli) [inline src/utils/WatsonUtils.ts]: modify comment: upload always a single workspace

File: src/utils/WatsonUtils.ts [hunk 3]
```diff
@@ -103,6 +104,20 @@ export function getWorkspaceInformationById(
...
+export function uploadWorkspaceInformationById(
+  watsonAssistantClient: watson.AssistantV1,
+  workspace: Workspace
```
Comment (CarmenManzulli) [inline src/utils/WatsonUtils.ts]: change type to UpdateWorkspace

File: src/utils/WatsonUtils.ts [hunk 4]
```diff
@@ -103,6 +104,20 @@ export function getWorkspaceInformationById(
...
+): Promise<Either<Error, Workspace>> {
+  logger.info(`Retrieving Information about a workspace Id`);
```
Comment (CarmenManzulli) [inline src/utils/WatsonUtils.ts]: Change log

File: src/utils/WatsonUtils.ts [hunk 5]
```diff
@@ -103,6 +104,20 @@ export function getWorkspaceInformationById(
...
+  return new Promise<Either<Error, Workspace>>(resolve => {
+    watsonAssistantClient.updateWorkspace(workspace, (err, response) => {
+      resolve(decodeWatsonResponse(err, response));
+      console.log(JSON.stringify(response));
```
Comment (CarmenManzulli) [inline src/utils/WatsonUtils.ts]: delete console.log(here) if you want to see the response use console.log(response) before resolve function

File: src/index.ts [hunk 6]
```diff
@@ -49,6 +49,29 @@ export async function startTool(): Promise<void> {
   }
   const watsonSourceClient = watsonSourceClientOrError.value;
 
+  // Retrieve a client for Target DB services
```
Comment (CarmenManzulli) [inline src/index.ts]: change comment into "Create db client for target"

File: src/index.ts [hunk 7]
```diff
@@ -49,6 +49,29 @@ export async function startTool(): Promise<void> {
...
+  const dbClientTargetOrError = DbUtils.getDb2Client(config.SOURCE.DB);
```
Comment (CarmenManzulli) [inline src/index.ts]: USE config.TARGET

File: src/index.ts [hunk 8]
```diff
@@ -49,6 +49,29 @@ export async function startTool(): Promise<void> {
...
+  if (isLeft(dbClientSourceOrError)) {
```
Comment (CarmenManzulli) [inline src/index.ts]: use the correct name: dbClientTargetOrError

File: src/index.ts [hunk 9]
```diff
@@ -49,6 +49,29 @@ export async function startTool(): Promise<void> {
...
+    logger.info("wrong result from getDb2Target");
+    endProcessHandler(undefined);
+    return;
+  }
+
```
Comment (CarmenManzulli) [inline src/index.ts]: delete this empty line

File: src/index.ts [hunk 10]
```diff
@@ -49,6 +49,29 @@ export async function startTool(): Promise<void> {
...
+  const dbClientTarget = dbClientTargetOrError.value;
+  // Catch if administrator is stopping the Server to avoid to keep open the DB connection
+  process.stdin.resume();
+  process.on(`SIGINT`, () => {
+    endProcessHandler(dbClientTarget);
+  });
```
Comment (CarmenManzulli) [inline src/index.ts]: modify endProcessHandler in order to use both dbClientSource and dbClientTarget

File: src/index.ts [hunk 11]
```diff
@@ -49,6 +49,29 @@ export async function startTool(): Promise<void> {
...
+  // CREATE CLIENT WATSON TARGET
+  const watsonTargetClientOrError = WatsonUtils.getWatsonAssistantClient(
+    config.TARGET.WATSON_API
+  );
+  if (isLeft(watsonTargetClientOrError)) {
+    console.log("error in target watson client");
+    return;
+  }
+
```
Comment (CarmenManzulli) [inline src/index.ts]: remember to insert const watsonTargetClient = const watsonTargetClientOrError.value

File: src/Configuration.ts [hunk 12]
```diff
@@ -54,7 +54,7 @@ export const Configuration = {
   MIGRATION_TOOL_PARAMETERS: {
```
Comment (CarmenManzulli) [inline src/Configuration.ts]: Add MIGRATION_TOOL_PARAMETERS on ReadMe and so for other missing configuration properties

File: src/utils/MigrationUtils.ts [hunk 13]
```diff
@@ -152,6 +152,9 @@ export async function getWorkspacesToMigrate(
     return left(dbWorkspacesOrError.value);
   }
   const dbWorkspaces = dbWorkspacesOrError.value;
+  if (dbWorkspaces.length === 0) {
+    throw Error("There aren't elements in database source");
```
Comment (CarmenManzulli) [inline src/utils/MigrationUtils.ts]: outside the try block you cannot use throw! Use return left() ...

File: src/utils/MigrationUtils.ts [hunk 14]
```diff
@@ -265,28 +267,21 @@ export async function uploadWorkspaces(
...
-          );
-
           const params = {
             workspace_id: targetWorkspacesId[0].workspaceId as string,
             name: workspaceToMigrate.workspace.name,
             description: workspaceToMigrate.workspace.description,
             language: workspaceToMigrate.workspace.language,
             intents: workspaceToMigrate.workspace.intents,
-            entities: workspaceToMigrate.workspace.entities,
+            entities: workspaceToMigrate.workspace.entities.slice(0, 15),
```
Comment (CarmenManzulli) [inline src/utils/MigrationUtils.ts]: remove slice()

File: src/utils/MigrationUtils.ts [hunk 15]
```diff
@@ -265,28 +267,21 @@ export async function uploadWorkspaces(
...
             dialog_nodes: workspaceToMigrate.workspace.dialog_nodes,
             counterexamples: workspaceToMigrate.workspace.counterexamples,
             metadata: workspaceToMigrate.workspace.metadata,
             learning_opt_out: workspaceToMigrate.workspace.learning_opt_out,
             system_settings: workspaceToMigrate.workspace.system_settings
-          } as UpdateWorkspaceParams;
-          console.log("intents converted");
+          } as any;
```
Comment (CarmenManzulli) [inline src/utils/MigrationUtils.ts]: remove as any, if you don't specify anything it will be any by default

File: src/utils/MigrationUtils.ts [hunk 16]
```diff
@@ -265,28 +267,21 @@ export async function uploadWorkspaces(
               `Non single result for ${workspaceToMigrate.dbWorkspaceName}`
             );
           }
-
-          console.log(
-            "\n\n\n\n-----------------------------------------------------------------------\n\n\n\n intent convertrtttttttttttttt"
-          );
-
           const params = {
```
Comment (CarmenManzulli) [inline src/utils/MigrationUtils.ts]: Create a function in order to build params

File: src/utils/MigrationUtils.ts [hunk 17]
```diff
@@ -297,7 +292,6 @@ export async function uploadWorkspaces(
             throw Error(errMsg);
           }
           const workspaceUpdate = workspaceUpdateOrError.value;
```
Comment (CarmenManzulli) [inline src/utils/MigrationUtils.ts]: directly return workspaceUpdateOrError.value

File: README.md [hunk 18]
```diff
@@ -27,23 +27,25 @@ It clones Workspaces from a source env to a target env filtering by Workspace Na
...
+| TARGET_DB2_DBNAME      | The DB2 DBNAME of Target Environment                          | string  | BLUDB      |
+| TARGET_DB2_HOSTNAME    | The DB2 Hostname of Target Environment                        | string  |            |
+| TARGET_DB2_UID         | The DB2 UID of Target Environment                             | string  |            |
+| TARGET_DB2_PWD         | The DB2 PWD of Target Environment                             | string  |            |
+| TARGET_DB2_PORT        | The DB2 PORT of Target Environment                            | number  |            |
+| TARGET_WATSON_USERNAME | The Watson Username of Target Environment                     | string  |            |
+| TARGET_WATSON_VERSION  | The Watson Version of Target Environment                      | string  | 2018-07-10 |
+| TARGET_WATSON_PASSWORD | The Watson Password of Target Environment                     | string  |            |
+| TARGET_WATSON_URL      | The Watson Url of Target Environment                          | string  |            |
+| MIGRATE_ALL            | All Workspace Id to Migrate if Migrate is Set to true         | boolean | true       |
```
Comment (CarmenManzulli) [inline README.md]: Comment is wrong! change it into: Migration tool flag - if true it migrates each workspace from source to target

File: README.md [hunk 19]
```diff
@@ -27,23 +27,25 @@ It clones Workspaces from a source env to a target env filtering by Workspace Na
...
+| SINGLE_WORKSPACE_ID    | The Single Workspace Id to Migrate if Migrate is Set to False | string  |            |
```
Comment (CarmenManzulli) [inline README.md]: Comment is wrong! change it into: Migration tool id - indicates the single workspace to migrate if MIGRATE_ALL is setted to true

File: src/Configuration.ts [hunk 20]
```diff
@@ -52,9 +52,9 @@ export const Configuration = {
     }
   },
   MIGRATION_TOOL_PARAMETERS: {
-    MIGRATE_ALL: false,
+    MIGRATE_ALL: process.env.MIGRATE_ALL || true,
     SINGLE_WORKSPACE_ID:
-      process.env.SINGLE_WORKSPACE_ID || "6423ca79-57d8-4305-8826-073e2ead33c3"
+      process.env.SINGLE_WORKSPACE_ID || "7cfe35a9-f710-4ddb-87bd-9015782aaa72"
```
Comment (CarmenManzulli) [inline src/Configuration.ts]: chose just one configuration!
- or migrate all in true with empty string for SINGLE_WORKSPACE_ID
- or migrate all is false and there is a non empty string for SINGLE_WORKSPACE_ID

File: src/index.ts [hunk 21]
```diff
@@ -17,7 +17,7 @@ export async function startTool(): Promise<void> {
   // Retrieve server configuration
   const configOrError = Configuration.getAppConfiguration();
   if (isLeft(configOrError)) {
-    logger.info("wrong result from getAppConfiguration");
+    logger.error("wrong result from getAppConfiguration");
```
Comment (CarmenManzulli) [inline src/index.ts]: change error message in each of these: not wrong result .... but: "Error getting App Configuration ${configOrError.value}" in this case. Use the same structure in other messages

File: src/index.ts [hunk 22]
```diff
@@ -56,7 +56,7 @@ export async function startTool(): Promise<void> {
     config.TARGET.WATSON_API
   );
   if (isLeft(watsonTargetClientOrError)) {
-    logger.info("error in target watson client");
+    logger.error("error in target watson client");
```
Comment (CarmenManzulli) [inline src/index.ts]: specify if source or target!!!! in each case

File: src/utils/WatsonUtils.ts [hunk 23]
```diff
@@ -106,7 +104,7 @@ export function getWorkspaceInformationById(
   });
 }
 
-// upload always a single workspace
+// upload a workspace
```
Comment (CarmenManzulli) [inline src/utils/WatsonUtils.ts]: change upload into update!

File: src/index.ts (hunk 22 above)
Comment (kaderson1) [reply src/index.ts]: viene già specificato se è source o target nel logger logger.info("**error in target watson client**");
non ho capito cosa volevi praticamente

Comment (CarmenManzulli) [reply src/index.ts]: Controlla ogni log in modo da renderlo chiaro e parlante....e magari aggiungici la visualizzazione dell'errore

File: src/index.ts [hunk 24]
```diff
@@ -39,16 +39,18 @@ export async function startTool(): Promise<void> {
...
   }
   const watsonSourceClient = watsonSourceClientOrError.value;
 
-  // Retrieve a client for Target DB services
+  // Create db client for target
   const dbClientTargetOrError = DbUtils.getDb2Client(config.TARGET.DB);
   if (isLeft(dbClientTargetOrError)) {
     logger.error(
-      `wrong result from getDb2Client target ${dbClientTargetOrError}`
+      `Error getting App Configuration Target ${dbClientTargetOrError}`
```
Comment (CarmenManzulli) [inline src/index.ts]: wrong message! it is the error getting Db2 client target!

File: src/index.ts [hunk 25]
```diff
@@ -39,16 +39,18 @@ export async function startTool(): Promise<void> {
     config.SOURCE.WATSON_API
   );
   if (isLeft(watsonSourceClientOrError)) {
-    logger.error(`error in source watson client ${watsonSourceClientOrError}`);
+    logger.error(
+      `Error getting Watson Asistant Client Source ${watsonSourceClientOrError}`
```
Comment (CarmenManzulli) [inline src/index.ts]: wrong word! it is Assistant

File: src/index.ts [hunk 26]
```diff
@@ -23,11 +23,11 @@ export async function startTool(): Promise<void> {
 
   const config = configOrError.value;
 
-  // Retrieve a client for Source DB services
+  // Create db client for source
   const dbClientSourceOrError = DbUtils.getDb2Client(config.SOURCE.DB);
   if (isLeft(dbClientSourceOrError)) {
     logger.error(
-      `wrong result from getDb2Client source ${dbClientSourceOrError}`
+      `Error getting App Configuration Source ${dbClientSourceOrError}`
```
Comment (CarmenManzulli) [inline src/index.ts]: wrong message! it is getting db2 client source!

File: src/index.ts [hunk 27]
```diff
@@ -60,7 +62,9 @@ export async function startTool(): Promise<void> {
     config.TARGET.WATSON_API
   );
   if (isLeft(watsonTargetClientOrError)) {
-    logger.error("error in target watson client");
+    logger.error(
+      `Error getting Watson Asistant Client Target ${watsonTargetClientOrError}`
```
Comment (CarmenManzulli) [inline src/index.ts]: againt....it is watson Assistant

File: src/utils/MigrationUtils.ts [hunk 28]
```diff
@@ -148,7 +148,9 @@ export async function getWorkspacesToMigrate(
     configMigrationParameters
   );
   if (isLeft(dbWorkspacesOrError)) {
-    logger.error(`wrong result from database source ${dbWorkspacesOrError}`);
+    logger.error(
+      `Error getting App Configuration from Database Source ${dbWorkspacesOrError}`
```
Comment (CarmenManzulli) [inline src/utils/MigrationUtils.ts]: wrong message!

**Pull request merged**